CONFIGURATION.md

```diff
@@ -0,0 +1,282 @@
+
+This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format. 
+
+> Notes:
+> - The validation is done during `HelixConfig.parse()`. 
+> - Using a IDE plugin, the yaml can also be validated during input.
+
+## Proposal: Strains are the only top-level concept
+
+This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.
+
+There will be no reconciliation of strains and some other concept at runtime. Every .hlx/strains.json will contain a fully deterministic, fully resolved and denormalized representation of all strains.
+
+## Proposal: Other complex configuration objects can be defined in a definitions container and re-used
+
+In order to make writing configurations easier and to enable re-use of configuration settings, we introduce a definitions container at the top level of the `helix-config.yaml` that can hold re-usable definitions like this:
+
+```yaml
+definitions:
+  strains:
+    base: &basestrain
+      code: /acapt/default/https---github-com-adobe-project-helix-io-git--master--
+      sticky: false
+    
+  origins:
+    publish: &publish
+      address: 192.168.0.1
+      name: publish
+      use_ssl: false
+```
+
+## Proposal: Configuration objects can be referenced using YAML references
+
+In order to re-use a complex configuration object (like an origin, a mount mapping, or even a strain), use YAML references (`*publish`) or YAML extensions (`<< *basestrain`). This allows us to rely on the behavior of the YAML parser instead of using home-grown inheritance and referencing logic that is likely less robust and certainly less well documented.
+
+A strain re-using the `*basestrain` above would look like this:
+
+```yaml
+  client:
+    <<: *basestrain
+    condition: req.http.host == "client.project-helix.io"
+    content:
+      repo: helix-cli
+      ref: master
+      owner: adobe
+    directoryIndex: readme.html
+```
+
+A strain referencing the `*publish` origin would look like this:
+```yaml
+  proxy-detailed:
+    sticky: true
+    origin: *publish
+```
+
+## Proposal: The `default` strain serves all traffic in production when no other strain is selected, therefore a `default` strain is mandatory
+
+We need one strain that is serving traffic that no other strain claims. This could be one by:
+
+1. requiring the existence of a `default` strain
+2. requiring the existence of one and only one strain that has the `default: true` property set
+
+Option (1) is easier to validate than option (2) and less error prone (every YAML editor will tell you if you have two keys called `default`), so I'd opt for (1)
+
+`hlx publish` should fail when no `default` strain exists.
+`hlx up` can continue without a `default` strain, but should issue a warning
```

Comment:
I think the `default` strain should be mandatory by schema. If we only allow no config or a valid config, we can provide a built-in default config that is used for no-config.